File: apps/desktop/src/main/services/fleetPermissions.ts
import { configService } from './configService';
import { claudeSessionStore } from './claudeSessionStore';

/** User-selected provider approval policy, never a facade/token grant.
 * Read live config for each launch so a running manager's next dispatch sees
 * settings changes. Follow recorded lineage only, with a cycle guard.
 */
export function fleetSkipsPermissions(
  opts: { manager?: boolean; parentSessionId?: string },
  enabled = configService.getConfig().agents?.fleetFullAccess === true,
): boolean {
  if (!enabled) return false;
  if (opts.manager) return true;
  const seen = new Set<string>();
  let parentId = opts.parentSessionId;
  while (parentId && !seen.has(parentId)) {
    seen.add(parentId);
    const parent = claudeSessionStore.getSnapshot(parentId);
    if (!parent) return false;
    if (parent.isWakeTarget || parent.isFleetManager) return true;
    parentId = parent.parentSessionId;
  }
  return false;
}

File: services/hub/cmd/brain/fleetpermissions.go
package main

// Fleet full access is a provider approval preference, never a token grant.
// Consult current config for every launch and follow recorded parent lineage.
func (r *registry) fleetSkipsPermissions(p spawnParams) bool {
	agents, _ := r.cfg.get()["agents"].(map[string]any)
	enabled, _ := agents["fleetFullAccess"].(bool)
	if !enabled {
		return false
	}
	if p.Manager {
		return true
	}
	seen := map[string]bool{}
	for id := p.ParentSessionID; id != "" && !seen[id] && r.meta != nil; {
		seen[id] = true
		parent, ok := r.meta.get(id)
		if !ok {
			return false
		}
		if parent.IsWakeTarget {
			return true
		}
		id = parent.ParentSessionID
	}
	return false
}
